File: 2020/8/both.pl
#!/usr/bin/perl

use strict;
use integer;

my $fh;

die "* no input\n" if (!open($fh, '<', $ARGV[0] || 'input.txt'));

my $linenum = 0;

my @prog;

while (1)
{
	my $line = <$fh>;
	$linenum++;
	last if (!$line);
	if ($line =~ /^(jmp|acc|nop) ([-+]\d+)/)
	{
		push @prog, [ $1, $2 ];
	}
}
my @res = run_prog(\@prog);
printf("Part 1 answer: %d\n", $res[1]);

for (my $pos = 0; $pos < @prog; $pos++)
{
	switch(\@prog, $pos);
	@res = run_prog(\@prog);
	if ($res[0] == @prog)
	{
		# run_prog(\@prog, 1);
		# printf("%d: -> %d, acc %d\n", $pos + 1, $res[0], $res[1]);
		last;
	}
	switch(\@prog, $pos);
}

printf("Part 2 answer: %d\n", $res[1]);

sub run_prog
{
	my ($prog, $trace) = @_;
	my $acc = 0;
	my $pos = 0;
	my @visited;

	while (!$visited[$pos] && ($pos != @$prog))
	{
		my $instr = $prog->[$pos];
		printf("%d: %s %s // acc %d`\n", $pos, $instr->[0], $pos + $instr->[1], $acc) if ($trace);
		$visited[$pos] = 1;
		if ($instr->[0] eq 'jmp')
		{
			$pos += $instr->[1];
		}
		elsif ($instr->[0] eq 'acc')
		{
			$acc += $instr->[1];
			$pos++;
		}
		elsif ($instr->[0] eq 'nop')
		{
			$pos++;
		}
		else
		{
			die "* Bad instruction at $pos\n";
		}
	}

	return ($pos, $acc);
}

sub switch
{
	my ($prog, $pos) = @_;
	my $instr = $prog->[$pos];
	if ($instr->[0] eq 'jmp')
	{
		$instr->[0] = 'nop';
	}
	elsif ($instr->[0] eq 'nop')
	{
		$instr->[0] = 'jmp';
	}
}

File: 2020/8/both.py
#!/usr/bin/python

import sys;
import re;

infile = 'input.txt'
if (len(sys.argv) > 1):
	infile = sys.argv[1]

fh = open(infile, 'r');

def run_prog(prog, trace = False):
	acc = 0
	pos = 0
	visited = [False] * len(prog)

	while pos < len(prog) and not visited[pos]:
		instr = prog[pos]
		if trace:
			print("%d: %s %s // acc %d`" % (pos, instr[0], pos + instr[1], acc))
		visited[pos] = True
		if instr[0] == 'jmp':
			pos += instr[1]
		elif instr[0] == 'acc':
			acc += instr[1]
			pos += 1
		elif instr[0] == 'nop':
			pos += 1
		else:
			raise Exception("bad instruction at %d" % pos)
	return (pos, acc)
	
prog = []

instr_re = re.compile(r'(jmp|acc|nop) ([-+]\d+)')
for line in fh:
	m = instr_re.match(line)
	if m:
		prog.append((m.group(1), int(m.group(2))))

res = run_prog(prog)

print("Part 1 answer: %d" % res[1])

def switch(prog, pos):
	instr = prog[pos]
	if instr[0] == 'jmp':
		prog[pos] = ('nop', instr[1]) 
	elif instr[0] == 'nop':
		prog[pos] = ('jmp', instr[1]) 

for pos in range(len(prog)):
	switch(prog, pos)
	res = run_prog(prog)
	if res[0] == len(prog):
		break
	switch(prog, pos)

print("Part 2 answer: %d" % res[1])
